### PR TITLE
implement restartless sync

### PR DESCRIPTION
**Pull request by diracdeltas**
fix https://github.com/brave/sync/issues/43

Test Plan:
1. open a pyramid
2. go to preferences, enable sync
3. go to some site, bookmark it. you should see it being synced in the console.
4. disable sync.
5. go to another site, bookmark it. you should see nothing in the console.
6. re-enable sync. you should see the site from Step 5 being synced.

### CODE REVIEW

**Comment by diracdeltas** [comment]
going to move Sync into a separate extension so https://github.com/brave/muon/pull/151 is not necessary

File: app/extensions.js [hunk 1]
```diff
@@ -368,7 +379,20 @@ module.exports.init = () => {
...
+    }
+    for (let i = 0; i < diff.length; i++) {
+      if (diff[i].path === '/settings/sync.enabled') {
+        return true
+      }
+    }
+    return false
+  }
+
+  let registerComponents = (diff) => {
```
Comment (ayumi) [inline app/extensions.js]: where is `diff` passed into this?

File: app/sync.js [hunk 2]
```diff
@@ -282,26 +290,39 @@ module.exports.init = function (initialState) {
...
-    return
-  }
-  ipcMain.on(messages.GET_INIT_DATA, (e) => {
-    const seed = initialState.seed || null
-    deviceId = initialState.deviceId || null
-    e.sender.send(messages.GOT_INIT_DATA, seed, deviceId, config)
-  })
   ipcMain.on(messages.SYNC_READY, module.exports.onSyncReady.bind(null,
-    !initialState.seed && !initialState.deviceId))
+    hasRun ? false : !initialState.seed && !initialState.deviceId))
```
Comment (ayumi) [inline app/sync.js]: minor: it's a little hard to read this ternary within a function call; might be cleaner to assign a variable right before.

File: app/extensions.js (hunk 1 above)
Comment (diracdeltas) [inline app/extensions.js]: `emitChanges` in js/stores/appStore

File: js/state/syncUtil.js [hunk 3]
```diff
@@ -168,6 +168,7 @@ const applySiteSettingRecord = (record) => {
  * @param {Object} record
  */
 module.exports.applySyncRecord = (record) => {
+  if (!record || !record.objectData) { return }
```
Comment (ayumi) [inline js/state/syncUtil.js]: minor: i think this should emit a warning

**Comment by ayumi** [comment]
looks sweet.

i haven't tried the Test Plan yet. should i wait for Sync as an Extension?

**Comment by diracdeltas** [comment]
@ayumi: addressed comments & made it its own extension

File: app/extensions.js [hunk 4]
```diff
@@ -214,6 +213,38 @@ let generateTorrentManifest = () => {
...
+    name: 'Brave Sync',
+    manifest_version: 2,
+    version: '1.0',
+    content_security_policy: concatCSP(cspDirectives),
+    content_scripts: [],
+    background: {
+      scripts: [ 'content/scripts/sync.js' ]
+    },
+    incognito: 'spanning',
+    key: 'MIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEAxOmBmOVzntEY6zrcrGSAyrhzL2FJt4FaP12nb899+SrV0LgpOgyqDjytuXT5IlHS74j7ZK2zTOTQy5/E9hqo6ioi1GA3PQU8E71DTaN6kW+XzP+VyZmgPoQHIxPg8jkYk/H4erfP9kMhkVOtu/XqDTqluNhOT0BvVlBpWd4unTQFWdgpCYlPrI6PsYya4FSuIDe6rCKtJABfuKFEr7U9d9MNAOJEnRS8vdBHWCuhWHqsfAaAPyKHQhnwFSFZ4eB+JznBQf7cQtB3EpOoBElyR9QvmbWFrYu87eGL5XxsojKHCrxlQ4X5ANsALa1Mdd2DHDMVqLMIiEEU42DVB0ZDewIDAQAB'
```
Comment (diracdeltas) [inline app/extensions.js]: the priv key for this is in 1password but i dont think anyone will ever need it

File: js/components/main.js [hunk 5]
```diff
@@ -456,6 +456,13 @@ class Main extends ImmutableComponent {
       windowActions.setImportBrowserDataSelected({})
     })
 
+    ipc.on(messages.SYNC_RELOADED, (e) => {
```
Comment (diracdeltas) [inline js/components/main.js]: oops, this is obsolete. removing

File: app/extensions.js [hunk 6]
```diff
@@ -214,6 +213,38 @@ let generateTorrentManifest = () => {
...
+
+  return {
+    name: 'Brave Sync',
+    manifest_version: 2,
+    version: '1.0',
+    content_security_policy: concatCSP(cspDirectives),
+    content_scripts: [],
+    background: {
+      scripts: [ 'content/scripts/sync.js' ]
+    },
```
Comment (ayumi) [inline app/extensions.js]: lets add some
```
icons: {
  128: 'img/sync-128.png',
  48: 'img/sync-48.png',
  16: 'img/sync-16.png'
},
```
sync-128
sync-48
sync-16

File: app/extensions.js [hunk 7]
```diff
@@ -300,6 +331,12 @@ module.exports.init = () => {
   process.on('extension-unloaded', (extensionId) => {
     extensionInfo.setState(extensionId, extensionStates.DISABLED)
     extensionActions.extensionDisabled(extensionId)
+    if (extensionId === config.syncExtensionId) {
+      // Reload sync extension to restart the background script
+      setImmediate(() => {
+        enableExtension(config.syncExtensionId)
```
Comment (diracdeltas) [inline app/extensions.js]: you can disable it functionally, but you can't disable the extension entirely; the latter isn't necessary afaict.

File: app/extensions.js (hunk 6 above)
Comment (diracdeltas) [inline app/extensions.js]: kk

**Pull request merged**